h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -1596,16 +1700,51 @@ private void doCleanup() {
     private transient Cholesky _chol;
     private transient L1Solver _lslvr;
 
+    private double[] constraintGLM_solve(GramGrad  gram) {
+      if (!_parms._intercept) throw H2O.unimpl();
+      ArrayList<Integer> ignoredCols = new ArrayList<>();
+      double[] xy = gram._xy.clone();
+      Cholesky chol = ((_state._iter == 0) ? gram.qrCholesky(ignoredCols, GamUtils.copy2DArray(gram._gram), _parms._standardize) : gram.cholesky(null, gram._gram));
+      if (!chol.isSPD()) throw new NonSPDMatrixException();
+      if (!ignoredCols.isEmpty()) {
+        int[] collinear_cols = ignoredCols.stream().mapToInt(x -> x).toArray();
+        String[] ignoredConstraints = collinearInConstraints(ArrayUtils.select(_dinfo.coefNames(), collinear_cols),
+                _state._csGLMState._constraintNames);
+        String collinear_col_names = Arrays.toString(ArrayUtils.select(_dinfo.coefNames(), collinear_cols));
+        if (ignoredConstraints != null && ignoredConstraints.length > 0)
+          throw new IllegalArgumentException("Found constraints " + Arrays.toString(ignoredConstraints) + 
+                  " included on collinear columns that are going to be removed.  Please remove any constraints " +
+                  "involving collinear columns.");
+        if (!_parms._remove_collinear_columns)
+          throw new Gram.CollinearColumnsException("Found collinear columns in the dataset. Set " +
+                  "remove_collinear_columns flag to true to remove collinear columns automatically. " +
+                  "Found collinear columns " + collinear_col_names);
+        _model.addWarning("Removed collinear columns "+collinear_col_names);
+        Log.warn("Removed collinear columns "+collinear_col_names);
+        _state.removeCols(collinear_cols);
+        gram._gram = GramGrad.dropCols(collinear_cols, gram._gram);
+        gram._grad = ArrayUtils.removeIds(gram._grad, collinear_cols);
+        xy = ArrayUtils.removeIds(xy, collinear_cols);
+      }
+      _chol = chol;
+      xy = xy.clone();
+      chol.solve(xy);
+      return xy;
+    }
+    
     private double[] ADMM_solve(Gram gram, double[] xy) {
-      if (_parms._remove_collinear_columns || _parms._compute_p_values) {
+      if (_state._noReg) {
```

Comment:
Since
```java
  // set _noReg to true if l1 and l2 regularization are both zero.
  public void setNoReg() {
    _noReg = (_lambda == 0);
  }
```

does not appear to reflect `_parms._remove_collinear_columns || _parms._compute_p_values` it looks to me like this might be a bug.

Reply:
You are right.  Fixed